Prohibit using Tensor rvalues with TensorExpressions

macedo22 opened the pull request:
## Proposed changes

This PR enforces prohibiting writing expressions that construct `Tensor`s inline, like:
- `Tensor<double> L = TensorExpressions::evaluate(Tensor<double>{1.0}() + R());`
- `Tensor<double> L = TensorExpressions::evaluate(S(ti_A, ti_a) + Tensor<double>{1.0}());`

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

macedo22 committed:
Prohibit using Tensor rvalues with TensorExpressions

wthrowe commented:
I haven't been following the TensorExpression stuff, so this may be naive, but this sounds like it would also forbid
```
Tensor<double> L = TensorExpressions::evaluate(function_that_returns_a_tensor()() + R());
```
Is that intended?

nilsdeppe commented:
Yea, the reason being that adding the logic to support rvalue cases makes memory usage a bit complicated. Basically, the expression representation of a Tensor is a combination of the tensor (effectively the Structure) plus the indices, e.g. `T(a,b)`. In order to avoid a lot of memory copies the `TensorAsExpression`, which is the representation of `T(a,b)`, holds a pointer to `T`. I think that pointer will dangle because the rvalue object that is bound to `const &` of the `operator+` is freed upon `operator+` returning but the `TensorExpression` representing the addition that `operator+` returned still needs to access that memory. Please correctly if I've got the object lifetime wrong here, because that would change what should be done in this PR.

Having said that, it would be possible to have `TensorAsExpression` either hold the data or pointer depending on whether you get an rvalue or an lvalue, but this also means that all the `operator+` etc. need to behave in a way that they use move semantics for their arguments. I think this is doable, but would lead to a somewhat unusual implementation (all `operator`s would need to take rvalues isn't what I've seen in other expression template libraries, but for e.g. vectors you don't have this late-time type information injection of the indices `a, b`). In short, I think to support rvalues correctly we would:
- need to forbid copying of expressions to avoid unintentional memory copies
- make all `operator`s only take rvalues (I'm not sure what happens when you take an rvalue to an abstract base class, actually)
- have `TensorAsExpression` be aware of whether it was generated through a `const&` or an `&&`
- have `operator()` in `Tensor` pass the required info to `TensorAsExpression`

macedo22 commented:
Thanks @wthrowe for bringing up that case - I hadn't thought about that. And thanks @nilsdeppe for the nice explanation. My understanding of the object lifetime is the same as yours based on experimenting with supporting doing operations with `DataVector` rvalues, but I'm happy to do a concrete test to confirm this. Let me know!

Regarding your bullet points on what would be required to support `Tensor` rvalues, I'm also in agreement with these points based on my attempt to support `DataVector` rvalues (without unnecessary copies) the last couple weeks with the `ScalarDataType` work I sent you and that we discussed today. In case we wanted to support `Tensor` rvalues, some notes on these points from this `ScalarDataType` work that I wanted to mention:
- *need to forbid copying of expressions to avoid unintentional memory copies*
I was able to eliminate *existing* copies with [these changes](https://github.com/sxs-collaboration/spectre/compare/develop...macedo22:move-rvalue-expressions). However, I didn't take the extra step of deleting the copy constructors for each of the expression classes simply because they would be required if we wanted to support doing operations on lvalue expressions. I think if we took the work I did and deleted copy constructors, this work would be done, we just wouldn't have support for doing operations on lvalue expressions without the copy constructors. And if we did want to support doing operations on lvalue expressions initially constructed from `Tensor` or `DataVector` rvalues, e,g,
```
auto lhs_expression_1 = DataVector{1.0, 2.0, 3.0} + R();
auto lhs_expression_2 = lhs_expression_1 + T();
```
then we have the implementation I came up with for `DataVector` rvalues that we discussed today that avoids copying the original `DataVector`, but, as I mentioned, the implementation for handling this felt a bit risky, and there is probably a better way.  This gist of how I implemented it is: in constructing `lhs_expression_2`, there would just be a pointer to the `DataVector` stored/owned by `lhs_expression_1`, and what makes the implementation risky is if there are copies happening when I think moves are happening and vice versa. And maybe we really don't care about having the ability to work with expression lvalues. I'm happy to break this down further with you later, if you'd like, over video chat or something.
- *make all operators only take rvalues (I'm not sure what happens when you take an rvalue to an abstract base class, actually)*
In the same branch linked above, I was able to get operators accepting rvalues working, even with the abstract base class, so that's good news!
- *have `TensorAsExpression` be aware of whether it was generated through a `const&` or an `&&`*
I didn't do this for `TensorAsExpression`, but I think we could do something similiar to the design for the combined `ScalarDataType` with a member pointer that either points internally to the owned value (moved rvalue `Tensor`) or externally to the `Tensor` that outlives the expression (what it currently does).

wthrowe commented:
> Yea, the reason being that adding the logic to support rvalue cases makes memory usage a bit complicated. Basically, the expression representation of a Tensor is a combination of the tensor (effectively the Structure) plus the indices, e.g. `T(a,b)`. In order to avoid a lot of memory copies the `TensorAsExpression`, which is the representation of `T(a,b)`, holds a pointer to `T`. I think that pointer will dangle because the rvalue object that is bound to `const &` of the `operator+` is freed upon `operator+` returning but the `TensorExpression` representing the addition that `operator+` returned still needs to access that memory. Please correctly if I've got the object lifetime wrong here, because that would change what should be done in this PR.

Temporaries live until the end of the full expression that lexically contains their creation (except for lifetime extension from references and some weird cases involving c arrays).  So arguments to functions can safely be referenced in their return values, but temporaries introduced inside the functions cannot.  (See, for example, `const T& std::max(const T& a const T& b)`.)  Pointer-vs.-reference doesn't matter here.

Is there a reason to deviate from the semantics Blaze uses for its expression templates?  Having different rules for what is valid depending on what type of expression template is in use sounds like a potential source of confusion.  In Blaze:
```c++
DataVector x{1.0};
DataVector from_expression = x + DataVector{1.0};  // fine                      
auto expr_no_temp = x + x;
DataVector from_no_temp = expr_no_temp;  // also fine                           
auto expr_temp = x + DataVector{1.0};
DataVector from_temp = expr_temp;  // BAD (but compiles)
```

nilsdeppe commented:
Ah, I clearly was misunderstanding that! @macedo22 could you add some print statements to destructors to verify that we don't have something weird going on with nested experssions where object lifetime is changed? Somethig like:

```cpp
T() = S(a) * F(A)+G()+H();
```
where you construct the Tensors inline or have them be returned from a function (basically, making sure they are rvalue references).

macedo22 commented:
@nilsdeppe @wthrowe Just tested this out with a commit on this [test-tensor-r-value](https://github.com/macedo22/spectre/tree/test-tensor-rvalue) branch. I left more notes in the comments in the tests. but here is a summary of the cases tested with helper functions that returned `Tensor`s with data type `DataVector`:
1. Create RHS expression using a temporary `Tensor`, assign it to a named LHS `Tensor`, then pass the named expression to evaluate ([link to test code](https://github.com/macedo22/spectre/blob/5a25d4983eca7e7669503bc254adbaf9a369a63f/tests/Unit/DataStructures/Tensor/Expressions/Test_TensorRValue.cpp#L95)):
```
  const auto expr_with_temp_tensor = get_rank2_tensor<DataVector>()(ti_a, ti_b);
  const auto actual_result_tensor_from_lvalue_expr =
      TensorExpressions::evaluate<ti_a, ti_b>(expr_with_temp_tensor);
```
**Result:** the temporary `Tensor` is destroyed in first line's assignment, so `expr_with_temp_tensor` is incorrect for evaluation in the next line. `CHECK`s fail.

2. Create RHS expression using a temporary `Tensor` and immediately pass the temporary expression to evaluate ([link to test code](https://github.com/macedo22/spectre/blob/5a25d4983eca7e7669503bc254adbaf9a369a63f/tests/Unit/DataStructures/Tensor/Expressions/Test_TensorRValue.cpp#L121)):
```
auto actual_result_tensor_from_rvalue_expr =
      TensorExpressions::evaluate<ti_a, ti_b>(
          get_rank2_tensor<DataVector>()(ti_a, ti_b));
```
**Result:** the temporary `Tensor` is destroyed after LHS `Tensor` components are computed with `evaluate`. `CHECK`s pass.

3. Create RHS expression using nested temporary `Tensor`s and immediately pass the temporary expression to evaluate ([link to test code](https://github.com/macedo22/spectre/blob/5a25d4983eca7e7669503bc254adbaf9a369a63f/tests/Unit/DataStructures/Tensor/Expressions/Test_TensorRValue.cpp#L197)):
```
const auto actual_result_tensor_from_rvalue_expr =
      TensorExpressions::evaluate<ti_a>(
          get_rank1_lower_tensor<DataVector>()(ti_a) +
          get_rank2_tensor<DataVector>()(ti_a, ti_b) *
              get_rank1_upper_tensor<DataVector>()(ti_B) +
          get_rank1_lower_tensor<DataVector>()(ti_a));
```
**Result:** the temporary `Tensor`s are destroyed after LHS `Tensor` components are computed with `evaluate`. `CHECK`s pass.

@nilsdeppe Let me know if this has implications for the `ScalarDataType`, because it seems like it may be possible to support temporary `DataVector`s as operands in TEs (unless we don't want that still, I think I recall you saying they are expensive?) with having it use the same pattern as `TensorAsExpression` (pointer to external object). We just won't be able to support the case like [this code snippet from above](https://github.com/sxs-collaboration/spectre/pull/2809#issuecomment-769592498) - essentially, the reuse of an expression constructed from temporary `DataVector`. If we don't care about being able to do something like this, then, I think this could work. Also, it sounds like @wthrowe was saying that doing this kind of thing with `DataVector`s (constructing a `DataVector` from a temporary, then constructing another `DataVector` from that - the `BAD (but compiles)` case [here](https://github.com/sxs-collaboration/spectre/pull/2809#issuecomment-770285830)) is bad, so maybe it should just follow that doing something similar with TEs (constructing a TE with a temporary `Tensor`, then constructing a new TE from the first TE) is also bad, and we don't do it.

nilsdeppe commented:
:+1: okay so then I think we can continue to allow r-values in the expressions. There aren't anymore unsafe things here than in e.g. Blaze or other expression template libraries.

I'm still undecided on allowing temporary `DataVector`s in expressions, largely for the inefficiency reasons. @wthrowe this is talking about allowing:
```cpp
auto result = evaluate<a>(DataVector{3.0, 7} * S(a))
```
What definitely needs to be allowed is:
```cpp
auto result evaluate<a>(3.0 * S(a))
```
The `DataVector` case is less clear since it has the same number at each grid point, and even if you were to have different numbers at each grid point, that smells like a bug because the user has assumed both a data layout and a mesh size. Do you think we should allow the `DataVector{...} * S(a)`-type expressions?

wthrowe commented:
Is there a reason to allow multiplying by DataVectors at all?  (Temporary or not?)  Shouldn't those be Scalars?

nilsdeppe commented:
The only reason would be to put them on similar footing as a `double`. I've been on the side of "While I could see someone wanting it, I would prefer not to be able to multiple by a `DataVector`". If you agree, let's go with not supporting it :)

macedo22 commented:
@wthrowe @nilsdeppe Are we in favor of closing this PR since it appears `Tensor` rvalues seem to work? I assume I'll be closing this, but wasn't sure if we were still trying to first confirm here, as part of the above thread, to not support `DataVector * TensorExpression`

nilsdeppe commented:
- I'm in favor of closing this PR 
- I think we shouldn't allow `DataVector * TensorExpression`, but `double * TensorExpression` is fairly commonly needed. @wthrowe what do you think?

wthrowe commented:
I agree.

macedo22 commented:
Sounds good! Thanks @wthrowe for bringing up the rvalue case I hadn't thought of, as well as the comments on the lifetime of temporaries in expressions
